Feature/code search tool

pranath-reddy opened the pull request:
### Summary :memo:
This PR adds Initial implementation of two code search tools.
- A semantic code search tool for querying local repositories using embedding-based similarity.
- A GitHub code search tool that wraps the SearxNG tool to query public GitHub repositories via keyword-based search.

### Details

#### `LocalRepoCodeSearchTool` (Semantic Search over Local Repositories)
- Introduced `LocalRepoCodeSearchTool` under `akd.tools.code_search` to perform semantic search over local repositories.
- Uses `Embedder` from `akd.tools.misc` to compute the embeddings.
- Automatically downloads `.csv` data file from Google Drive if missing.
- Adheres to the AKD tool format.

#### `GitHubCodeSearchTool` (Keyword-Based Search via SearxNG)
- Configures a `SearxNGSearchTool` instance restricted to the GitHub engine and `technology` category.
- Allows querying GitHub repositories via keyword search using the endpoint [search/repositories](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-repositories).
- Included `examples/code_search_test.py` to test tools with custom queries asynchronously.

### Usage

#### LocalRepoCodeSearchTool

```python
from akd.tools.code_search import (
    LocalRepoCodeSearchToolConfig,
    CodeSearchToolInputSchema,
    LocalRepoCodeSearchTool
)

# Configure the tool
search_cfg = LocalRepoCodeSearchToolConfig(debug=True)

# Initialize the local semantic code search tool
search_tool = LocalRepoCodeSearchTool(config=search_cfg)

# Run the semantic search
result = await search_tool._arun(
    CodeSearchToolInputSchema(
        queries=["landslide nepal"],
        max_results=5
    )
)
```

#### GitHubCodeSearchTool

```python
from akd.tools.code_search import (
    GitHubCodeSearchTool,
    CodeSearchToolInputSchema
)
from akd.tools.search import (
    SearxNGSearchToolConfig
)

# Configure the GitHub code search tool
search_cfg = SearxNGSearchToolConfig(score_cutoff=0.1)

# Initialize the GitHub search tool
search_tool = GitHubCodeSearchTool(config=search_cfg)

# Run the GitHub code search
result = await search_tool._arun(
    CodeSearchToolInputSchema(
        queries=["landslide nepal"],
        max_results=10
    )
)
```

### Bugfixes :bug: (delete if dind't have any)
-

### Checks
- [ ] Closed #798
- [ ] Tested Changes
- [ ] Stakeholder Approval

pranath-reddy committed:
Add semantic code search tool

This commit introduces a reusable component for performing semantic searches on a code repository dataset.

pranath-reddy committed:
Add specialized GitHubSearchTool

Create a new GitHubSearchTool as a wrapper around the SearxNGSearchTool

pranath-reddy committed:
Update code_search_test.py

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+    content: str = Field(
+        ...,
+        description="A snippet of content or description from the repository.",
+    )
```
Can we reuse the `akd.structures.SearchResultItem` here? Because technically it'd be the same? If not: move this to `akd.structures.CodeSearchResultItem` which directly inherits from `SearchResultItem`.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+from akd.tools._base import BaseTool, BaseToolConfig
+
+from akd.tools.gcd.embeddings import Embedder
+from akd.tools.gcd.vectorizer import RepoFinder
```
Let's move gcd to `akd.tools.misc` subpackage as the usages of these are just internally exposed

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+from akd.tools.gcd.vectorizer import RepoFinder
+from akd.tools.search import SearxNGSearchTool, SearxNGSearchToolConfig, SearxNGSearchToolInputSchema, SearxNGSearchToolOutputSchema
+
+class CodeSearchToolInputSchema(InputSchema):
```
Maybe 
```python

class CodeSearchToolInputSchema(SearchToolInputSchema):
...
```

would be better inheritance pattern.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+        description="A snippet of content or description from the repository.",
+    )
+
+class CodeSearchOutputSchema(OutputSchema):
```
Maybe we can reuse `SearchToolOutputSchema` here?

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+
+class CodeSearchToolConfig(BaseToolConfig):
+    """Configuration for the repository search tool."""
+    data_file: str = os.getenv("CODE_SEARCH_DATA_FILE", ".code_data/repositories_with_embeddings.csv") # TODO: Replace with SDE API in future.
```
make the `.code_data/` folder to `docs/` directly for now and reference it via akd root path.

There's `akd.utils.get_akd_root` function that reutns the root path.

For reference, you can check `akd.tools.source_validator.SourceValidatorConfig.whitelist_file_path`

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,242 @@
+        """
+        config = config or self.config_schema()
+        super().__init__(config, debug)
+        self.repo_finder = None
```
can we consolidate repo finder to this class itself? the logic. And we don't have to maintain 2 different things in akd. Have most of the logic as repo finder, but no need to have repo finder as separate thing. 

And Have CodeSearchTool as a base type. and have `LocalCodeRepoCodeSearchTool` or something.


```python
class CodeSearchTool(BaseTool[CodeSearchToolInputSchema, CodeSearchOutputSchema]):
    """
    Tool for performing semantic code search. 
    It automatically downloads the necessary data file if it's not found locally.
    """
    input_schema = CodeSearchToolInputSchema
    output_schema = CodeSearchOutputSchema
    config_schema = CodeSearchToolConfig

    def __init__(self, config: CodeSearchToolConfig | None = None, debug: bool = False):
    ...

class LocalRepoCodeSearchTool(...):
  ...
```

This way, when we build more code search tool that's better, we can just have it typed to `CodeSearchTool`

pranath-reddy replied:
Will add the base `CodeSearchTool` after I get the SDE code search tool ready. The other changes are done.

pranath-reddy committed:
Addressed PR feedback

- Worked on feedback from Nish
- merged Repo Finder logic with main tool
- Test script updated at examples/code_search_test

pranath-reddy committed:
Update input schema

- Input schema for local repo search now inherits from SearchToolInputSchema

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,291 @@
+        description="The maximum number of repository results to return.",
+    )
+
+class LocalRepoCodeSearchToolConfig(BaseToolConfig):
```
Can we do this:
Have `CodeSearchTool` as a base class and inherit any downstream codesearch tool from it? It will help keep the semantics better because we will have multiple codesearch tool. This also avoids polluting the upper hierarchy like `SearchTool`.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,291 @@
+                os.makedirs(data_dir, exist_ok=True)
+            
+            # Download from Google Drive
+            file_id = "1nPaEWD9Wuf115aEmqJQusCvJlPc7AP7O"
```
We need to put this file information in config. We don't want to hardcode it. Have it in config.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,291 @@
+            
+            # Download from Google Drive
+            file_id = "1nPaEWD9Wuf115aEmqJQusCvJlPc7AP7O"
+            gdown.download(f"https://drive.google.com/uc?id={file_id}", data_file_path, quiet=False)
```
Put the google drive download thing in `akd.utils` for separation of concerns.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,291 @@
+Tool for performing targeted searches on GitHub using SearxNG.
+This tool is a wrapper around SearxNGSearchTool.
+"""
+class GitHubCodeSearchTool(SearxNGSearchTool):
```
Once we have `CodeSearchTool` as a bae, we can also do multiple inheritance for this. In python the resolution happens from right to left manner for multiple inheritance.

eg:
```python
class GitHubCodeSearchTool(CodeSearchTool, SearxNGSearchTool)
...
```
```

pranath-reddy committed:
Addressed PR Feedback

- Moved drive file ID to config
- Moved file download logic to akd.utils
- Added abstract base class for code search tools

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,317 @@
+            
+            # Download from Google Drive
+            file_id = self.config.google_drive_file_id
+            google_drive_downloader(file_id, data_file_path, quiet=False)
```
can you add logger.debug mode statement if debug is turned on to make sure why the initialization is taking longer when I initialize the code search tool.

NISH1001 replied:
If it's there, it's fine. Just making a comment.

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,317 @@
+    """
+    Input schema for the local repository code search tool.
+    """
+    top_k: int = Field(
```
can we make this top_k as `computed_field` and return `max_results`?

```python
from pydantic import computed_field

....
@computed_field
def top_k(self) -> int:
    return self.max_results
```

pranath-reddy committed:
Update pyproject.toml

- Add gdown dependency

pranath-reddy committed:
Update pyproject.toml

- Add pandas dependency

pranath-reddy committed:
Addressing PR comments

- added debug statements and made top_k a computed field

NISH1001 committed:
Validate IO for code search with multiple inheritance

Override _validate_input and _validate_output in `CodeSearchTool`.

NISH1001 committed:
Merge branch 'feature/code_search_tool' of github.com:NASA-IMPACT/accelerated-discovery into feature/code_search_tool

Fix merge conflicts

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,395 @@
+    output_schema = CodeSearchToolOutputSchema
+    config_schema = CodeSearchToolConfig
+
+    def _validate_input(
+        self,
+        params: CodeSearchToolInputSchema | SearchToolInputSchema | dict,
+    ) -> CodeSearchToolInputSchema:
```
This allows us to have any input semantics as long as they are codesearch input schema or just search input schema

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -0,0 +1,395 @@
+            )
+        return params
+
+    def _validate_output(
+        self,
+        output: CodeSearchToolOutputSchema | SearchToolOutputSchema,
+    ) -> CodeSearchToolOutputSchema:
```
Here we validate the output and type check if it also is searrchtooloutput schema because in the `_arun` we're calling `super()` which gives output of type `SearxNGSearchToolOutputSchema` which is semantically the same as any search tool output schema
